### ISSUE TITLE
[Feature request]: Support for toggling between Light and Dark polarity

### ISSUE THREAD

**Issue by TheColorman** [danth/stylix#447]
I like synchronizing my system color to the sunset and sunrise, but that's not possible with Stylix unless I do a nixos rebuild. I'd love for Stylix to generate both a dark and a light polarity theme on rebuild, for the OS to then be able to switch between them using a program like [baduhai/Koi](https://github.com/baduhai/Koi) or [oskarsh/Yin-Yang](https://github.com/oskarsh/Yin-Yang) (or even built in to Stylix itself?)

Manual/dynamic switching between light and dark themes is increasingly popular, so I think Stylix supporting it would make a lot of sense, though I'm not sure how much effort it would take to implement such a feature.

**Comment by danth** [comment]
This is difficult to implement since only a very small number of apps can switch based on a toggle (such as in GNOME), and only about half of apps have automatic reload if we were to change their config on a timer.

**Comment by trueNAHO** [comment]
https://github.com/danth/stylix/issues/63 could be extended in the future to change wallpapers with a certain color palette based on the time.

**Comment by BhasherBEL** [comment]
One way to achieve it, at least for the apps that works with "hot reload", could be to use [specialisation](https://nixos.wiki/wiki/Specialisation). A very basic example would be:
```nix
    stylix = {
      enable = true;
      image = ./assets/mountains_dark.jpg;
      polarity = "dark";
      base16Scheme = "${pkgs.base16-schemes}/share/themes/catppuccin-macchiato.yaml";
    };

    specialisation.day.configuration = {
      stylix = {
        image = lib.mkForce ./assets/mountains_light.jpg;
        polarity = lib.mkForce "light";
        base16Scheme = lib.mkForce "${pkgs.base16-schemes}/share/themes/catppuccin-latte.yaml";
      };
    };
```

**Comment by trueNAHO** [comment]
> One way to achieve it, at least for the apps that works with "hot reload", could be to use [specialisation](https://nixos.wiki/wiki/Specialisation).

Feel free to submit a PR adding this to [`docs/src/tricks.md`](https://github.com/danth/stylix/blob/94d70292d0c687ebacb65d00bd516cbefa18d3ca/docs/src/tricks.md).

**Comment by no-mood** [comment]
> One way to achieve it, at least for the apps that works with "hot reload", could be to use [specialisation](https://nixos.wiki/wiki/Specialisation). A very basic example would be:
> 
> ```nix
>     stylix = {
>       enable = true;
>       image = ./assets/mountains_dark.jpg;
>       polarity = "dark";
>       base16Scheme = "${pkgs.base16-schemes}/share/themes/catppuccin-macchiato.yaml";
>     };
> 
>     specialisation.day.configuration = {
>       stylix = {
>         image = lib.mkForce ./assets/mountains_light.jpg;
>         polarity = lib.mkForce "light";
>         base16Scheme = lib.mkForce "${pkgs.base16-schemes}/share/themes/catppuccin-latte.yaml";
>       };
>     };
> ```

Do specialisation work for you? Can't get the "hot reload" as many things (including GTK apps) don't switch unless i close and re-open them

**Comment by trueNAHO** [comment]
> Do specialisation work for you? Can't get the "hot reload" as many things (including GTK apps) don't switch unless i close and re-open them

Hot reload support is application dependant. See https://github.com/danth/stylix/discussions/371 for more information.

**Comment by TheColorman** [comment]
@BhasherBEL just tried it out, seems to work pretty nicely! Do Stylix options such as `config.lib.stylix.colors` also get updated when they are used outside of the specialisation?

**Comment by BhasherBEL** [comment]
> @BhasherBEL just tried it out, seems to work pretty nicely! Do Stylix options such as `config.lib.stylix.colors` also get updated when they are used outside of the specialisation?

Specialisations can be used multiple ways, but literally build a new system for every specialisation, so yeah, it should not be a problem.

**Comment by TheColorman** [comment]
If anyone's interested I've written a small service that uses [heliocron](https://github.com/mfreeborn/heliocron) to automatically switch polarity on sunset/sunrise. It can probably be improved, but it does the job: [`TheColorman/nixcfg/modules/profiles/stylix/default.nix`](https://github.com/TheColorman/nixcfg/blob/e31e3ec6db8f3f0b64ab9a917b623b676bd2b97f/modules/profiles/stylix/default.nix)

I'm still not too happy about having to do an entire `nixos-rebuild` just to switch theme, as it feels a bit overkill, but this works for now.

**Comment by trueNAHO** [comment]
> If anyone's interested I've written a small service that uses [heliocron](https://github.com/mfreeborn/heliocron) to automatically switch polarity on sunset/sunrise. It can probably be improved, but it does the job: [`TheColorman/nixcfg/modules/profiles/stylix/default.nix`](https://github.com/TheColorman/nixcfg/blob/e31e3ec6db8f3f0b64ab9a917b623b676bd2b97f/modules/profiles/stylix/default.nix)

This is a great PoC. Thanks for sharing!

> I'm still not too happy about having to do an entire `nixos-rebuild` just to switch theme, as it feels a bit overkill, but this works for now.

Aside from replacing `nixos-rebuild` with Nix specialisations and replacing the service loop on success with a simple infinite Bash loop to improve logging, this looks mostly good.

**Comment by TheColorman** [comment]
> Aside from replacing nixos-rebuild with Nix specialisations and replacing the service loop on success with a simple infinite Bash loop to improve logging, this looks mostly good.

Issue with specialisations is that specialisations can't refer to each other, so assuming there was both a light and a dark specialisation, the `/run/current-system/specialisation/` directory is empty after you switch to one of the specialisations, and the only way to switch to the other specialisations is a full `nixos-rebuild`, apparently due to some issues with infinite recursive references.

As far as I understand, activating the specialisation through `/run/current-system/specialisation/` is functionally the same as running `nixos-rebuild test --specialisation`.

**Comment by noahfraiture** [comment]
Hello, is there any news since the POC ? It looks great and it would be really neat !
In the mean time, except for the time needed, is there disadvantage to rebuild ? If it upgrade isn't it possible to disable the upgrade to be sure to make the minimal number of new store ?

**Comment by danth** [comment]
> In the mean time, except for the time needed, is there disadvantage to rebuild?

Some targets don't reload their theme until the app is restarted, so the new theme may not be applied consistently throughout the system. This is something we cannot fix without modifying the apps themselves to support hot reloading.

> If it upgrade isn't it possible to disable the upgrade to be sure to make the minimal number of new store?

If you use specialisations, these are pre-built, so only the activation stage of the rebuild actually happens at the time you apply the specialisation. This means that switching to a specialisation won't pull in package updates.

**Comment by noahfraiture** [comment]
> > In the mean time, except for the time needed, is there disadvantage to rebuild?
> 
> Some targets don't reload their theme until the app is restarted, so the new theme may not be applied consistently throughout the system. This is something we cannot fix without modifying the apps themselves to support hot reloading.

Yes of course, but that's another problem

> 
> > If it upgrade isn't it possible to disable the upgrade to be sure to make the minimal number of new store?
> 
> If you use specialisations, these are pre-built, so only the activation stage of the rebuild actually happens at the time you apply the specialisation. This means that switching to a specialisation won't pull in package updates.

Oh thank you, I didn't know about specialisation, I'll search for more information about it

**Comment by Weathercold** [comment]
> > Aside from replacing nixos-rebuild with Nix specialisations and replacing the service loop on success with a simple infinite Bash loop to improve logging, this looks mostly good.
> 
> Issue with specialisations is that specialisations can't refer to each other, so assuming there was both a light and a dark specialisation, the `/run/current-system/specialisation/` directory is empty after you switch to one of the specialisations, and the only way to switch to the other specialisations is a full `nixos-rebuild`, apparently due to some issues with infinite recursive references.
> 
> As far as I understand, activating the specialisation through `/run/current-system/specialisation/` is functionally the same as running `nixos-rebuild test --specialisation`.

You can still find the specialisations by searching. I wrote a darkman module to switch polarity: https://github.com/Weathercold/nixfiles/blob/master/home/modules/services/scheduling/darkman.nix

Although that script stops working if the specialisation gets GC'ed

**Comment by husjon** [comment]
I recently starting moving over to stylix and found this to be one feature that I really missed from my previous setup.
I use darkman to switch polarity based on time of day.

Since I already used and have some darkman hooks set up for some of my applications I could just cycle darkman to to get the correct polarity when the state diverged.

I made a quick post-activation script using home-manager to do this automatically.
During the day (when light-mode should be active) and polarity is set to dark, it will still activate dark-mode, but once activation is done, cycle darkman to restore the state.

In case anyone also use darkman and home-manager you can find the post-activation script here
https://github.com/husjon/nixos-config/blob/main/modules/services/darkman/default.nix#L45-L57

Darkman hooks for various application that I use can be found at https://github.com/husjon/nixos-config/tree/main/modules/services/darkman

Note: I should add that this isn't a fix, rather a workaround that does not require a rebuild of the system for switching :)

**Comment by barsikus007** [comment]
specialisation workaround generate unnecessary boot entries
I made own workaround for that (will make nixpkgs PR soon)
https://github.com/barsikus007/config/commit/c8314942d96da9db03f01e6f4f9825483be8e6d2


also, made auto switching with darkman and sudo exclusions:
https://github.com/barsikus007/config/commit/4c95355c112480755bc40b07366b7776ddbcfc0d

**Comment by TheColorman** [comment]
> specialisation workaround generate unnecessary boot entries

Is there a benefit to using `switch` instead of `test` here? The only difference between `test` and `switch` is that [`switch` generates boot entries, whereas `test` doesn't](https://github.com/NixOS/nixpkgs/blob/c0b0e0fddf73fd517c3471e546c0df87a42d53f4/pkgs/by-name/sw/switch-to-configuration-ng/src/src/main.rs#L1094-L1098). You can use `test` with specialisations just fine.

EDIT: fixed the hyperlink, I had linked to the wrong source code.

**Comment by barsikus007** [comment]
> Is there a benefit to using `switch` instead of `test` here?

when you generate your usual configuration with switch next time, specialisation stay in your config and generate entry

**Comment by TheColorman** [comment]
Ah, so you want it to set your profile, but not generate a boot entry, is that correct? That makes sense.

**Comment by kim366** [comment]
I understand that not every app supports hot reload and I don't see it as the responsibility of stylix to implement this.

GTK and Qt do support hot reload, though. Making it possible to specify two themes per program (light and dark; if supported by that program) or light/dark variations inside one theme (I am unsure how Qt/GTK choose their other theme) would fully satisfy my usecase.

Is this already possible with stylix? What happens currently when the xdg-desktop-portal switches appearance? I don't yet use stylix, so I can't try it out

**Comment by TheColorman** [comment]
> Is this already possible with stylix? What happens currently when the xdg-desktop-portal switches appearance? I don't yet use stylix, so I can't try it out

The Stylix theme and polarity have a single value, and this value is part of the NixOS configuration in the Nix store. So to change between Dark and Light, you currently need to switch to a different NixOS system configuration which has your preferred theme.

The big change Stylix would have to make is move from an architecture where the theme gets computed and stored in the Nix store, to one where multiple themes can be present at the same time and switchable during runtime, which is pretty difficult right now.

Let's take Kitty as an example. Stylix uses Home Manager's Kitty module to apply the Kitty theme. Home Manager generates `~/.config/kitty/kitty.conf` on every rebuild/switch which contains a line that loads the correct Stylix color palette. How would Stylix manage this in runtime? Essentially, the `kitty.conf` file would need to be replaced whenever the theme changes, or the way the Kitty theme is loaded would have to drastically change. But Stylix has no control over this file - its managed entirely by Home Manager. Stylix just uses Home Manager to add some Kitty config.

Repeat this for every single application supported by Stylix, and this becomes a pretty complicated problem. If Stylix would need to be able to switch theme dynamically at runtime, it would have to essentially emulate a `nixos-rebuild switch`, only updating the files that contain Stylix theming.

Hence, most of the solutions in this thread discuss some way to automate running `nixos-rebuild switch` or similar with NixOS Specialisations.

**Comment by trueNAHO** [comment]
> I understand that not every app supports hot reload and I don't see it as the responsibility of stylix to implement this.
>
> GTK and Qt do support hot reload, though. Making it possible to specify two themes per program (light and dark; if supported by that program) or light/dark variations inside one theme (I am unsure how Qt/GTK choose their other theme) would fully satisfy my usecase.
>
> Is this already possible with stylix?

Yes, this is possible for applications supporting hot reloading: https://github.com/nix-community/stylix/issues/1863. This just requires upstreaming hot reload triggers.

> The big change Stylix would have to make is move from an architecture where the theme gets computed and stored in the Nix store, to one where multiple themes can be present at the same time and switchable during runtime, which is pretty difficult right now.

I suppose this is the exact use case of specialisations.

**Comment by husjon** [comment]
Going down the specialisation rabbit hole for theme switching sounds like a bad idea in my mind. One is as mentioned earlier due to it taking some time to perform the switching, however the idea is worse if you're already using specialisations for other things.

As far as I can tell there is no way of combining specialisations (as in switching to a combination of specialisations, i.e. `home` and `dark`).
As such if I already have 2 specialisations (`home` and `work`), I would need to make additional specialisations to cover each permutation (`home-dark`, `home-light`, `work-dark`, `work-light`). This would as you can hopefully tell get out of hand quickly if you need more variations.

**Comment by trueNAHO** [comment]
> Going down the specialisation rabbit hole for theme switching sounds like a bad idea in my mind. One is as mentioned earlier due to it taking some time to perform the switching

Directly running the generated `/run/current-system/specialisation/<SPECIALISATION>/activate` activation scripts is probably as fast as it gets without actively skipping critical work.

> however the idea is worse if you're already using specialisations for other things.
>
> As far as I can tell there is no way of combining specialisations (as in switching to a combination of specialisations, i.e. `home` and `dark`). As such if I already have 2 specialisations (`home` and `work`), I would need to make additional specialisations to cover each permutation (`home-dark`, `home-light`, `work-dark`, `work-light`). This would as you can hopefully tell get out of hand quickly if you need more variations.

This approach of permuting theme variants with existing specialisations, offloads specialisation declarations onto end-users. Alternatively, Stylix re-implements a variant of specialisations such that theme variants can be switched within the same configuration at run time:

> The big change Stylix would have to make is move from an architecture where the theme gets computed and stored in the Nix store, to one where multiple themes can be present at the same time and switchable during runtime, which is pretty difficult right now.

Since switching Stylix configuration generally requires running at least parts of the `activate` script for correctness, implementing a variant of specialisations should probably generate the complete `activate` script by permuting theme variants with existing specialisations. Essentially, this specialisation implementation would do what end-users would otherwise manually have to do, and then provides an interface for switching between Stylix specialisations.

**Comment by husjon** [comment]
> Directly running the generated `/run/current-system/specialisation/<SPECIALISATION>/activate` activation scripts is probably as fast as it gets without actively skipping critical work.

Using `activate` on my system yields no visual change at all, opening new windows etc still have the same polarity and theme as the default.

For reference I'm using this setup to configure the specialisations while testing this approach.
https://gist.github.com/husjon/200cc747d113f30f680e5e4c458d8f2c

Only when using `/nix/var/nix/profiles/system/specialisation/<SPECIALISATION>/bin/switch-to-configuration test/switch` does it actually change.


> Alternatively, Stylix re-implements a variant of specialisations such that theme variants can be switched within the same configuration at run time

I think this is a much more sane approach although I'm not sure how it would be implemented (user facing could be a helper command ala `stylix switch <polarity/theme>`), not entirely sure about backend.
This could also open up for allowing a user to predefine a couple of themes and not just polarities.

**Comment by barsikus007** [comment]
`/run/current-system/specialisation/<SPECIALISATION>/activate` script works if your apps is checking theme paths (ie, on my system with plasma, it reloads well)
also, some stylix theming, like fzf one, won't affect by activate script

**Comment by kim366** [comment]
GTK Themes like Adwaita-Dark no longer exist, because the dark variant is built into the main theme nowadays. I feel like this would be the simplest thing Stylix could do:

Instead of one polarity setting, have two theme definitions, one for dark, one for light, and have a preference setting. The programs that support it, get the merged theme of both and other programs build only the preferred theme ibto their config and the other theme separately into the Nix store.

It is then the responsibility of the user to do dynamic switching. I don't know about kitty, but Hyprland has source (like import/include) that I use for all sorts of such automated configuration switching while keeping the actual config file readonly

**Comment by Kernald** [comment]
I definitely agree that support for apps with support for both polarities (GTK, Ghostty...), this would be 1) much easier on the Stylix side of things, 2) likely a better user-experience in the long run.

**Comment by trueNAHO** [comment]
> > Alternatively, Stylix re-implements a variant of specialisations such that theme variants can be switched within the same configuration at run time
>
> [...] This could also open up for allowing a user to predefine a couple of themes and not just polarities.

Yes, this opens up new niche use cases, although polarity switching is probably the most relevant one.

> Instead of one polarity setting, have two theme definitions, one for dark, one for light, and have a preference setting. The programs that support it, get the merged theme of both and other programs build only the preferred theme ibto their config and the other theme separately into the Nix store.

This has been previously proposed in https://github.com/nix-community/stylix/pull/1899 and will not generally work when applications do not support this. The previously discussed Stylix specialisations might generally work, [to the extent hot reloading is possible](https://github.com/nix-community/stylix/issues/1863), and support any number of themes. The Stylix specialisation would also be a better approach than the previously very briefly discussed alternative of [declaring dark and light values for all Stylix options](https://github.com/nix-community/stylix/issues/1855).

**Comment by Kernald** [comment]
> The previously discussed Stylix specialisations might generally work, to the extent hot reloading is possible

But what about applications for which hot reloading is _not_ possible? In a world where applications are more and more day/night theme aware, it feels like not taking advantage of that but rather requiring a hook/signal/... for every application to reload its entire configuration to change its theme dynamically is the wrong way to approach this.

**Comment by trueNAHO** [comment]
> > The previously discussed Stylix specialisations might generally work, to the extent hot reloading is possible
>
> But what about applications for which hot reloading is _not_ possible? In a world where applications are more and more day/night theme aware, it feels like not taking advantage of that but rather requiring a hook/signal/... for every application to reload its entire configuration to change its theme dynamically is the wrong way to approach this.

I assume internal dark and light theme switching is primarily supported by GUI applications, while hot reloading probably covers more applications overall because it is generally easier to implement than an entire internal theme switching interface. I suggest implementing Stylix specialisations first because it probably does not require refactoring the entire Stylix interface and might have more impact with less effort.

**Comment by bbaserdem** [comment]
Been thinking about this issue lately; since I use stylix and been trying to utilize the polarity options in a few apps that I have that support light/dark mode switching (ghostty, kitty). I think this feature would need a significant refactor of stylix.

Hot-reloading should be out of scope for stylix. Apps should support hot reloading upon polarity change themselves; this should not be the job of stylix. The day/night switching of polarity likewise should be out of scope for stylix. That should remain a userspace tooling and DE feature. Stylix should implement theming, not patch behavior into apps that are not natively supported. Not to mention dynamic behavior is not really the domain of home-manager/nixos/nix-darwin; these tools are for static configuration. Specialization can work for now, but it's ridiculous to switch configurations daily to do theme toggling.

The way I think about this is; the following is needed for light and dark theming;

- Refactor config options that allows generating or defining a dark and a light theme
- Have an option that defines a default polarity to use as fallback for apps that don't support polarity features.
- Rewrite the modules for apps that allow for polarities to use both variants. (This would be the main painful refactor, as each modules' maintainer needs to fix their module to use the new refactored stylix.)
- For modules with programs and services that don't support polarity mode, use the fallback polarity as the default theme.

This would be a significant refactor of stylix; not sure if this is doable without breaking changes.

**Comment by trueNAHO** [comment]
> Hot-reloading should be out of scope for stylix. Apps should support hot reloading upon polarity change themselves; this should not be the job of stylix.

This is already the plan: https://github.com/nix-community/stylix/issues/1863.

> The day/night switching of polarity likewise should be out of scope for stylix.

Toggling between dark and light themes would be a special case of the previously discussed Stylix specialisations. Whether to implement this special case in Stylix or provide its implementation in `/doc/src/tricks.md` should be decided once the Stylix specialisation interface exists.

> That should remain a userspace tooling DE feature.

Stylix can be used as a userspace tool and Stylix should work standalone without explicit WM/DE support.

> Stylix should implement theming, not patch behavior into apps that are not natively supported.

Although I generally agree, this is already done to support certain environments.

> Not to mention dynamic behavior is not really the domain of home-manager/nixos/nix-darwin; these tools are for static configuration.

The current plan is to generate a static script to switch theme variants at run time:

> This approach of permuting theme variants with existing specialisations, offloads specialisation declarations onto end-users. Alternatively, Stylix re-implements a variant of specialisations such that theme variants can be switched within the same configuration at run time:
>
> > The big change Stylix would have to make is move from an architecture where the theme gets computed and stored in the Nix store, to one where multiple themes can be present at the same time and switchable during runtime, which is pretty difficult right now.
>
> -- https://github.com/nix-community/stylix/issues/447#issuecomment-3720838078

> Specialization can work for now, but it's ridiculous to switch configurations daily to do theme toggling.

Since the theme is coupled to the configuration for reproducibility, switching themes requires switching configurations.

> The way I think about this is; the following is needed for light and dark theming;
>
> - Refactor config options that allows generating or defining a dark and a light theme

As previously mentioned, Stylix specialisations would generally support any number of themes beyond only dark and light themes:

> The previously discussed Stylix specialisations might generally work, [...], and support any number of themes. The Stylix specialisation would also be a better approach than the previously very briefly discussed alternative of [declaring dark and light values for all Stylix options](https://github.com/nix-community/stylix/issues/1855).
>
> -- https://github.com/nix-community/stylix/issues/447#issuecomment-3724774425

> - Have an option that defines a default polarity to use as fallback for apps that don't support polarity features.
>
> - Rewrite the modules for apps that allow for polarities to use both variants. (This would be the main painful refactor, as each modules' maintainer needs to fix their module to use the new refactored stylix.)
>
> - For modules with programs and services that don't support polarity mode, use the fallback polarity as the default theme.
>
> This would be a significant refactor of stylix; not sure if this is doable without breaking changes.

As previously explained, it would be better to first implement Stylix specialisations:

> > > The previously discussed Stylix specialisations might generally work, to the extent hot reloading is possible
> >
> > But what about applications for which hot reloading is _not_ possible? In a world where applications are more and more day/night theme aware, it feels like not taking advantage of that but rather requiring a hook/signal/... for every application to reload its entire configuration to change its theme dynamically is the wrong way to approach this.
>
> I assume internal dark and light theme switching is primarily supported by GUI applications, while hot reloading probably covers more applications overall because it is generally easier to implement than an entire internal theme switching interface. I suggest implementing Stylix specialisations first because it probably does not require refactoring the entire Stylix interface and might have more impact with less effort.
>
> -- https://github.com/nix-community/stylix/issues/447#issuecomment-3729473208

**Comment by Sk7Str1p3** [comment]
this could be implemented by enabling stylix to generate palettes for both `light` and `dark` polarities (e.g by setting `stylix.polarity = "dynamic" # or "both"`), and link 2 versions of app config, and writing quite simple bash script for renaming files
